kzg/src/msm/bos_coster.rs

```diff
@@ -0,0 +1,217 @@
+                heap.push(Pair {
+                    scalar: *s,
+                    point_index: idx,
+                });
+            }
+        }
+
+        while heap.len() > 1 {
+            let Some(pair1) = heap.pop() else {
+                break;
+            };
+            let pair2_scalar = {
+                let Some(pair2) = heap.peek_mut() else {
+                    break;
+                };
+                // the optimization for large scalar differences was taken from: https://link.springer.com/article/10.1007/s13389-012-0027-1
+                let (k, pair2_scalar_2powk) = max_pow2_multiple_leq(&pair1.scalar, &pair2.scalar);
+                let mut new_point = work_points[pair1.point_index].clone();
+                for _ in 0..k {
+                    new_point.dbl_assign();
+                }
+
+                work_points[pair2.point_index].add_or_dbl_assign(&new_point);
+
+                pair2_scalar_2powk
+            };
+
+            let scalar = pair1.scalar.sub(pair2_scalar);
+            if !scalar.is_zero() {
+                heap.push(Pair {
+                    scalar,
+                    point_index: pair1.point_index,
+                });
+            }
+        }
+
+        if let Some(pair) = heap.pop() {
+            work_points[pair.point_index].mul(&TFr::from_u64_arr(&pair.scalar.data))
+        } else {
+            TG1::zero()
+        }
+    }
+}
+
+impl Ord for Scalar256 {
+    #[inline(always)]
+    fn cmp(&self, other: &Self) -> Ordering {
+        for i in (0..4).rev() {
+            if self.data[i] < other.data[i] {
+                return Ordering::Less;
+            } else if self.data[i] > other.data[i] {
+                return Ordering::Greater;
+            }
+        }
+        Ordering::Equal
+    }
+}
+
+trait LeadingZeros {
+    fn leading_zeros(&self) -> u32;
+}
+
+impl LeadingZeros for Scalar256 {
+    #[inline(always)]
+    fn leading_zeros(&self) -> u32 {
+        for i in (0..4).rev() {
+            if self.data[i] != 0 {
+                return (i as u32) * 64 + (64 - self.data[i].leading_zeros());
+            }
+        }
+        0
+    }
+}
+
+impl Shl<u32> for Scalar256 {
+    type Output = Scalar256;
+
+    // This function was generated with ChatGPT
+    #[inline(always)]
+    fn shl(self, n: u32) -> Scalar256 {
+        if n == 0 {
+            return self;
+        }
+        let limb_shift = (n / 64) as usize;
+        let bit_shift = n % 64;
+        let mut res = [0u64; 4];
+
+        for i in (0..4).rev() {
+            if i < limb_shift {
+                continue;
+            }
+            let mut val = self.data[i - limb_shift] << bit_shift;
+            if bit_shift != 0 && i - limb_shift > 0 {
+                val |= self.data[i - limb_shift - 1] >> (64 - bit_shift);
+            }
+            res[i] = val;
+        }
+        Scalar256::from_u64(res)
+    }
+}
+
+trait Shr1Assign {
+    fn shr_once_assign(&mut self);
```

Comment:
I don't think this is good idea to create custom method just for `>> 1` -- just implement general `Shr`, and with `inline(always)` compiler will optimize that usecase for you.